File: flutter/test/e2e/control.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:retry/retry.dart';

Map<String, String> headers = {'Content-Type': 'application/json'};

/// Client of a Control API.
class Client {
  http.Client inner = http.Client();
  late String controlApiAddress;

  /// Returns a new Control API [Client].
  Client(this.controlApiAddress);

  /// Creates the provided media `Element` in the provided [path] on a media
  /// server.
  Future<http.Response> create(String path, Object element) async {
    var response = await retry(() => inner.post(
        Uri.parse(getUrl(controlApiAddress, path)),
        headers: headers,
        body: json.encode(element)));
    if (response.statusCode != 200) {
      throw response.body;
    }
    return response;
  }

  /// Deletes a media `Element` identified by the provided [path].
  Future<http.Response> delete(String path) async {
    var response = await retry(
        () => inner.delete(Uri.parse(getUrl(controlApiAddress, path))));
    if (response.statusCode != 200) {
      throw response.body;
    }
    return response;
  }

  /// Returns a media `Element` identified by the provided [path].
  Future<http.Response> get(String path) async {
    var response = await retry(
        () => inner.get(Uri.parse(getUrl(controlApiAddress, path))));
    if (response.statusCode != 200) {
      throw response.body;
    }
    return response;
  }

  /// Applies on a media server the provided media `Element` identified by the
  /// provided [path].
  Future<http.Response> apply(String path, Object element) async {
    var response = await retry(() => inner.put(
        Uri.parse(getUrl(controlApiAddress, path)),
        headers: headers,
        body: json.encode(element)));
    if (response.statusCode != 200) {
      throw response.body;
    }
    return response;
  }

  /// Fetches all callbacks received by Control API mock server.
  Future<http.Response> callbacks() async {
    var response =
        await retry(() => inner.get(Uri.parse('$controlApiAddress/callbacks')));
    if (response.statusCode != 200) {
      throw response.body;
    }
    return response;
  }

  /// Returns URL of a media element identified by the provided [path].
  String getUrl(String controlApiAddress, String path) {
    return '$controlApiAddress/control-api/$path';
  }
}

File: e2e/tests/control.rs
//! HTTP client interacting with Medea via its Control API.

use derive_more::{Display, Error, From};
use medea_control_api_mock::{
    api::Response,
    callback::CallbackItem,
    proto::{CreateResponse, Element, SingleGetResponse},
};

/// All errors which can happen while working with a Control API.
#[derive(Debug, Display, Error, From)]
pub enum Error {
    Reqwest(reqwest::Error),
}

type Result<T> = std::result::Result<T, Error>;

/// Client of a Control API.
pub struct Client {
    inner: reqwest::Client,
    control_api_address: String,
}

impl Client {
    /// Returns a new Control API [`Client`].
    #[must_use]
    pub fn new(control_api_address: &str) -> Self {
        Self {
            inner: reqwest::Client::new(),
            control_api_address: control_api_address.to_owned(),
        }
    }

    /// Creates the provided media [`Element`] in the provided `path` on a Medea
    /// media server.
    pub async fn create(
        &self,
        path: &str,
        element: Element,
    ) -> Result<CreateResponse> {
        let gg : CreateResponse = self
        .inner
        .post(&get_url(&self.control_api_address, path))
        .json(&element)
        .send()
        .await?
        .json()
        .await?;
        println!("create {:?}", gg);
        if let Some(err) = gg.error {
            panic!();
        }
        Ok(gg)
    }

    /// Deletes a media [`Element`] identified by the provided `path`.
    pub async fn delete(&self, path: &str) -> Result<Response> {
        let gg: Response = self
        .inner
        .delete(&get_url(&self.control_api_address, path))
        .send()
        .await?
        .json()
        .await?;
        println!("delete {:?}", gg);
        if let Some(err) = gg.error {
            panic!();
        }
        Ok(gg)
    }

    /// Returns a media [`Element`] identified by the provided `path`.
    pub async fn get(&self, path: &str) -> Result<SingleGetResponse> {
        let gg: SingleGetResponse = self
        .inner
        .get(&get_url(&self.control_api_address, path))
        .send()
        .await?
        .json()
        .await?;
        println!("delete {:?}", gg);
        if let Some(err) = gg.error {
            panic!();
        }
        Ok(gg)
    }

    /// Applies on a media server the provided media [`Element`] identified by
    /// the provided `path`.
    pub async fn apply(
        &self,
        path: &str,
        element: Element,
    ) -> Result<CreateResponse> {
        let gg: CreateResponse = self
        .inner
        .put(&get_url(&self.control_api_address, path))
        .json(&element)
        .send()
        .await?
        .json()
        .await?;
        println!("apply {:?}", &gg);
        if let Some(err) = gg.error {
            panic!();
        }
        Ok(gg)
    }

    // TODO: Server side filtering on GET requests or SSE/WS subscription would
    //       speed up things. We a probably wasting a lot of time on ser/deser
    //       of huge JSON's.
    /// Fetches all callbacks received by Control API mock server.
    pub async fn callbacks(&self) -> Result<Vec<CallbackItem>> {

        let gg: Vec<CallbackItem> = self
        .inner
        .get(&format!("{}/callbacks", self.control_api_address))
        .send()
        .await?
        .json()
        .await?;
        
        Ok(gg)
    }
}

/// Returns URL of a media [`Element`] identified by the provided `path`.
fn get_url(control_api_address: &str, path: &str) -> String {
    format!("{control_api_address}/control-api/{path}")
}
